### PR TITLE
Jenkins-mage-package

### PR DESCRIPTION
**Pull request by DaveSys911**
Jenkins release pipeline v1.

Addressing tasks:
https://github.com/elastic/security-team/issues/3125
https://github.com/elastic/security-team/issues/3404
Multi branch jenkins pipeline that runs unit-tests, build, package & upload to GCS bucket and Artifacts API when running on main branch.

image

### CODE REVIEW

**Comment by tehilashn** [comment]
This requires a 🎉 !

File: .ci/jobs/cloudbeat-mbp.yml
Comment (oren-zohar) [inline .ci/jobs/cloudbeat-mbp.yml:16]: we can remove irrelevant versions from here

File: .ci/jobs/cloudbeat-mbp.yml
Comment (oren-zohar) [inline .ci/jobs/cloudbeat-mbp.yml:32]: we can remove irrelevant versions from here

File: .ci/jobs/cloudbeat-mbp.yml
Comment (oren-zohar) [inline .ci/jobs/cloudbeat-mbp.yml:33]: case-sensitive means anything in this context? these are numbers

File: .ci/jobs/cloudbeat-mbp.yml
Comment (oren-zohar) [inline .ci/jobs/cloudbeat-mbp.yml:36]: same as above

File: .ci/scripts/linux-test.sh
Comment (oren-zohar) [inline .ci/scripts/linux-test.sh:23]:
```suggestion
# Download systemtest dependencies so the download time doesn't count towards
```

File: .ci/scripts/package-docker-snapshot.sh
Comment (oren-zohar) [inline .ci/scripts/package-docker-snapshot.sh:3]: if this does the same as the `mage package` CI job we have we should consider disabling it

File: .ci/scripts/package-docker-snapshot.sh
Comment (oren-zohar) [inline .ci/scripts/package-docker-snapshot.sh:14]: conflicting with `package.sh` is it ok?

File: .ci/scripts/prepare-spec-changes.sh
Comment (oren-zohar) [inline .ci/scripts/prepare-spec-changes.sh:13]: not sure we have this files in our repo

File: .ci/scripts/rm-docker.sh
Comment (oren-zohar) [inline .ci/scripts/rm-docker.sh:20]: consider removing this part

File: .ci/scripts/rm-docker.sh
Comment (oren-zohar) [inline .ci/scripts/rm-docker.sh:4]: let's comment why this is hard coded

File: .ci/scripts/unit-test.sh
Comment (oren-zohar) [inline .ci/scripts/unit-test.sh:15]: consider removing this part

File: magefile.go
Comment (oren-zohar) [inline magefile.go:207]:
```suggestion
// PythonEnv ensures the Python venv is up-to-date with the beats requirements.txt.
```

File: scripts/make/check_changelogs.py
Comment (oren-zohar) [inline scripts/make/check_changelogs.py:8]: looks like our version is not supported so this script will do nothing?

File: utils/tools.go
Comment (oren-zohar) [inline utils/tools.go:1]: not urgent for now, but we should think of a way to consume this package instead of copying it.

File: scripts/make/generate_notice.py
Comment (oren-zohar) [inline scripts/make/generate_notice.py:18]: do we need windows here?

File: scripts/make/copy-docs.py
Comment (oren-zohar) [inline scripts/make/copy-docs.py:18]: I don't think we have these files in the repo, do you think this script will affect anything?

File: Jenkinsfile
Comment (oren-zohar) [inline Jenkinsfile:9]: are these creds temporary? or do we intend to use yours in the long run? can be problematic imo

File: Jenkinsfile
Comment (oren-zohar) [inline Jenkinsfile:117]: since we're running unit tests in the PRs CI we should consider removing one of them, since this is meant to run only post-merge I think we should remove this one as it's redundant, wdyt?

File: Jenkinsfile
Comment (oren-zohar) [inline Jenkinsfile:167]: on the other hand, this one should stay over our current CI package, right?

File: Makefile
Comment (oren-zohar) [inline Makefile:98]: can be removed, no x-pack here or planned to be afaik

File: Makefile
Comment (oren-zohar) [inline Makefile:59]: why?

File: Makefile
Comment (oren-zohar) [inline Makefile:1]: many of the commands here have mage alternative, we should consider replacing them in order to have better maintainability, this is not a blocker for this PR but I just want to document it:
test, clean, check, update, add-headers, and more.

**Review by oren-zohar — CHANGES_REQUESTED** [review]
Amazing job! 🚀  add some comments that we may want to address in the future, and some we can fix rn.

File: .ci/jobs/cloudbeat-mbp.yml
Comment (DaveSys911) [reply .ci/jobs/cloudbeat-mbp.yml:16]: Done

File: .ci/jobs/cloudbeat-mbp.yml
Comment (DaveSys911) [reply .ci/jobs/cloudbeat-mbp.yml:32]: Done

File: .ci/jobs/cloudbeat-mbp.yml
Comment (DaveSys911) [reply .ci/jobs/cloudbeat-mbp.yml:33]: It's a regex, not sure if it will match a suffix like "8.x.x-alpha". I prefer to keep it restrictive in any case and match branches exactly.

File: .ci/scripts/package-docker-snapshot.sh
Comment (DaveSys911) [reply .ci/scripts/package-docker-snapshot.sh:3]: These scripts are a wrapper around mage commands, to set env-vars/preconditions for mage itself(make release runs mage package).

It makes the pipeline easier to maintain.

File: .ci/scripts/rm-docker.sh
Comment (DaveSys911) [reply .ci/scripts/rm-docker.sh:20]: It will be used soon, it passes the branch param to the release manager docker - Since our branches won't  map out 1to1 to the gradle config files they use per project/branch.

File: .ci/scripts/unit-test.sh
Comment (DaveSys911) [reply .ci/scripts/unit-test.sh:15]: Regarding unit-tests, i prefer to keep them disabled and not removing them. Since we might need to run our prs on jenkins as well.

File: .ci/scripts/prepare-spec-changes.sh
Comment (DaveSys911) [reply .ci/scripts/prepare-spec-changes.sh:13]: removed

File: Jenkinsfile
Comment (DaveSys911) [reply Jenkinsfile:9]: Will be removed in a follow up pr, without these i have no abillity to test e2e locally. Can remove after i see main works successfully on the actual jenkins cluster

File: Makefile
Comment (DaveSys911) [reply Makefile:1]: I agree. From what i see in other repos. Tools like MAKE or JUST are used as wrappers around mage commands and service other functions mage lacks. They should compliment each other and not overlap, We will need to organize this point.

File: Jenkinsfile
Comment (DaveSys911) [reply Jenkinsfile:117]: I agree, set default to false-these will be skipped

File: Jenkinsfile
Comment (DaveSys911) [reply Jenkinsfile:167]: The script & make command are a wrapper around the mage package  command.It's used to set conditions/vars before invoking mage.

File: Makefile
Comment (DaveSys911) [reply Makefile:59]: The other commands are used for pr builds,update is the only one relevant for main branches I can remove that line - i suggest to clean it up in a batch.

File: Makefile
Comment (DaveSys911) [reply Makefile:98]: removed

File: scripts/make/check_changelogs.py
Comment (DaveSys911) [reply scripts/make/check_changelogs.py:8]: removed

File: scripts/make/copy-docs.py
Comment (DaveSys911) [reply scripts/make/copy-docs.py:18]: can be removed

File: scripts/make/generate_notice.py
Comment (DaveSys911) [reply scripts/make/generate_notice.py:18]: Prob not, but this script is required and i'd prefer to avoid changes.

File: utils/tools.go
Comment (DaveSys911) [reply utils/tools.go:1]: Depends on future pipeline iterations, If we keep it, worth a shot.

File: .ci/jobs/cloudbeat-mbp.yml
Comment (DaveSys911) [reply .ci/jobs/cloudbeat-mbp.yml:33]: done

File: .ci/scripts/rm-docker.sh
Comment (DaveSys911) [reply .ci/scripts/rm-docker.sh:4]: Done

File: Jenkinsfile
Comment (oren-zohar) [reply Jenkinsfile:9]: sounds good to me

File: Makefile
Comment (oren-zohar) [reply Makefile:59]: I'm not sure I understood your answer, maybe I wasn't clear with my comment. There are two `check-full` commands, one is commented out which I assume really checking the entire flow (`update`, `check`, `golint`, `staticcheck`, `check-docker`,`compose`) and the other one only runs `update`. Is this by design? from what I see it's being used in the intake file so I'm curious of why you made it check less stuff, does the other checks fail?

File: Jenkinsfile
Comment (oren-zohar) [reply Jenkinsfile:167]: not sure this is answering what I asked

**Review by oren-zohar — CHANGES_REQUESTED** [review]
Saw your comments and answered, there are still some comments that need to be addressed. Also, I see your CLA check is failing, you'll need to either rebase [f1945ed](https://github.com/elastic/cloudbeat/pull/16/commits/f1945ed8c7afa75bda509fc65c16e05041efa971), [cf1d50d](https://github.com/elastic/cloudbeat/pull/16/commits/cf1d50da38f40942892a8f88ddb66ed98f3f4461), [8b8d194](https://github.com/elastic/cloudbeat/pull/16/commits/8b8d1946288c5d28ee76452fbef564230532628f) or make sure your DaveOps account sign a [Contributor Agreement](https://www.elastic.co/contributor-agreement).

File: Makefile
Comment (DaveSys911) [reply Makefile:59]: Exactly, some failed others applied formatting to the code, and were intended for pr branches.So i left the minimum required for the snapshot build.

File: Jenkinsfile
Comment (DaveSys911) [reply Jenkinsfile:167]: If u mean the gh package ci job. Then no, it should stay in place. We aren't planning to run this workflow for pr checks and we want to know if 'mage package' fails before we merge a pr.

File: Makefile
Comment (oren-zohar) [reply Makefile:59]: okay if we have failing test we should address them, although this indeed not relevant to this PR we should document this failing test and add them to our CI

File: scripts/make/generate_notice.py
Comment (oren-zohar) [reply scripts/make/generate_notice.py:18]: alrighty

File: Jenkinsfile
Comment (oren-zohar) [reply Jenkinsfile:167]: alrighty

File: magefile.go
Comment (oren-zohar) [reply magefile.go:207]: the `requirements.txt` is empty though?

File: Makefile
Comment (DaveSys911) [reply Makefile:59]: These checks aren't mandatory and not necessarily have to be implemented exactly how they are done here. if we decide to do it, they need to go in our pr checks since it's mostly styling & formating.
The reason i disabled them, is because they will otherwise fmt/lint every file - which would've resulted in massive conflicts with main.

the checks are:

1. mage check-fmt & mage check-headers
2. honnef.co/go/utils/cmd/staticcheck
3. golint
4. docker integration test

File: .ci/scripts/package-docker-snapshot.sh
Comment (DaveSys911) [reply .ci/scripts/package-docker-snapshot.sh:14]: They are not conflicting, this one is a wrapper for mage package to build docker & package.sh is for binaries/tar.gz

File: .ci/scripts/unit-test.sh
Comment (DaveSys911) [reply .ci/scripts/unit-test.sh:15]: It's an opinionated approach to this. I think you need to differentiate between application code & build code when it comes to quality standards.

This commented-out code will be needed in the near future & it is tested and working.  The feedback loop on this bit is quite tedious. Since this script can only run on Jenkins due to vault permissions.

File: magefile.go
Comment (DaveSys911) [reply magefile.go:207]: The tooling fails without that function

**Review by oren-zohar — APPROVED** [review]
💯

**Pull request merged**
